### PR TITLE
Blogpost about how to use wireshark with protobuf and grpc dissectors

### PR DESCRIPTION
**Pull request by huangqiangxiong**
Add this blogpost as recommended on https://github.com/protocolbuffers/protobuf/issues/3303.

There may be some grammatical errors. You are welcome to modify them directly if you like.

Preview: https://deploy-preview-467--grpc-io.netlify.app/blog/how-to-analyze-grpc-with-wireshark/

### CODE REVIEW

**Comment by huangqiangxiong** [comment]
@jtattermusch hello, what else do I need to do about this PR? Or need I submit an issue for it?

**Comment by chalin** [comment]
Hi @huangqiangxiong - no, nothing more to be done. I've assigned @jtattermusch and @ejona86 as technical reviewers; if you think that someone else from the gRPC team might be a good/more suitable reviewer, let me know. I'll review (from a tech writers p.o.v), most likely early next week. /cc @thisisnotapril

**Comment by huangqiangxiong** [comment]
> Hi @huangqiangxiong - no, nothing more to be done. I've assigned @jtattermusch and @ejona86 as technical reviewers; if you think that someone else from the gRPC team might be a good/more suitable reviewer, let me know. I'll review (from a tech writers p.o.v), most likely early next week. /cc @thisisnotapril

OK. Thank you for your efforts.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+  google.protobuf.Timestamp last_updated = 5;
+  bytes portrait_image = 6;
+}
+
+message AddressBook {
+  repeated Person people = 1;
+}
+```
+
+According to the line `'import "google/protobuf/timestamp.proto"'`, the file `addressbook.proto` depends on the official *.proto library of Protobuf.
```
Comment (jtattermusch) [inline content/blog/wireshark-protobuf-and-grpc-dissectors.md:77]: the "official *.proto library of Protobuf"  is technically called "well known types", timestamp WKT is one of them.

https://developers.google.com/protocol-buffers/docs/reference/google.protobuf

**Review by jtattermusch — APPROVED** [review]
Thanks for the blogpost, that's a very useful reading and I like that there's a plenty of examples.

@chalin  can you please review from the techwriter perspective?

**Comment by chalin** [comment]
@jtattermusch - thanks for your review! I'll do mine shortly.

**Comment by chalin** [comment]
@ejona86 has somment comments about the content, so I'll wait for those before I make my pass.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md (hunk 1 above)
Comment (chalin) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md:77]: @huangqiangxiong, will you be making this change?

**Comment by chalin** [comment]
@huangqiangxiong - I have some general questions for you about this blog entry:

- What is your target audience?
- In two sentences or less, can you summarize the blog entry (so that, in reading your one or two sentences, someone would feel interested in reading the entire post :).
- After reading this blog entry, what should the reader have learned?
  That is, what are the _expected **learning outcomes**_?

(Once you've answered those questions, you might want to rethink the title.)

At the moment, the blog entry is feeling a little too **heavy on the details**.

You might consider dropping the following
(@ejona86 feel free to chime in):

- Details about the **JSON-serialized capture**. (I think that it is enough to mention support for JSON in the feature list)
- Details about **UDP**
- "_Write your own Protobuf UDP or TCP dissectors_" how often does a user have to write a custom dissector? If it isn't often, then consider removing the details from the text and simply referring the reader to Wireshark documentation.

I have other comments about the text, but I'll wait for some feedback from @ejona86.

I would like to point out one broad change that I feel will be necessary: "protobuf" isn't an English word and, as far as I know, it isn't an accepted abbreviation for "protocol buffers" or ".proto file". You should use the latter, as appropriate, instead.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md (hunk 1 above)
Comment (huangqiangxiong) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md:77]: @chalin, I was going to make this change after collecting comments from all three of you. Do you mean I can change it at any time?

**Comment by huangqiangxiong** [comment]
@chalin, thank you for your comments. Please see my reply.

> * What is your target audience?

I think the audience of this blog is mainly the following two kinds of people:
1. The people who want to debug or analyze their own gRPC or protobuf (Protocol Buffers) based messages transferred over the wire (network) by Wireshark.
2. The people who want to study the binary wire format of gRPC protocol or protocols based on protobuf. (Just as I wanted when I started learning protobuf and gRPC)

The original motivation of writing this article comes from the people's requirements from following links:
https://github.com/protocolbuffers/protobuf/issues/3303
https://github.com/grpc/grpc/issues/13586

> * In two sentences or less, can you summarize the blog entry (so that, in reading your one or two sentences, someone would feel interested in reading the entire post :).

With Wireshark, people can analyze the messages of gRPC protocol or protocols based on Protocol Buffers (a.k.a protobuf) that transferred over the wire (network), or learning the binary wire format of these protocols.

> * After reading this blog entry, what should the reader have learned?
>   That is, what are the _expected **learning outcomes**_?

This tutorial will show you how to configure and use Wireshark offical protobuf and gRPC dissectors to analyze grpc and protobuf based protocols with .proto files.

> (Once you've answered those questions, you might want to rethink the title.)

How about the title "How to dissect grpc and Protobuf based protocols with Wireshark"?

> At the moment, the blog entry is feeling a little too **heavy on the details**.
> 
> You might consider dropping the following
> (@ejona86 feel free to chime in):
> 
> * Details about the **JSON-serialized capture**. (I think that it is enough to mention support for JSON in the feature list)

OK, I will remove it.

> * Details about **UDP**
> * "_Write your own Protobuf UDP or TCP dissectors_" how often does a user have to write a custom dissector? If it isn't often, then consider removing the details from the text and simply referring the reader to Wireshark documentation.

Indeed, the two features of **UDP** and "_Write your own Protobuf UDP or TCP dissectors_" are only related to protobuf and have nothing to do with grpc. But they are important to Protobuf only users. If you don't think it's appropriate to put this content here, I will just simply leave a link referring to Wireshark documentation.

> I have other comments about the text, but I'll wait for some feedback from @ejona86.
> 
> I would like to point out one broad change that I feel will be necessary: "protobuf" isn't an English word and, as far as I know, it isn't an accepted abbreviation for "protocol buffers" or ".proto file". You should use the latter, as appropriate, instead.

But it seems that the word "protobuf" is also being used by the protocol buffers official code repository at https://github.com/protocolbuffers/protobuf. Or how about to use "Protocol Buffers" in the title or at the beginning of the article, like this: "Protocol Buffers (a.k.a., Protobuf)", and then still using "Protobuf" later?

**Comment by chalin** [comment]
@huangqiangxiong, thanks for your thoughtful reply and, in particular, for the link to the issues that provide the background / motivation for your blog post. I feel that we're making progress in clarifying intent and scope.

How about if we approached this incrementally? Could you make a pass to strip out the large chunks that we seem to be agreeing to remove: JSON-serialized capture, UDP and "Write your own..." (while leaving "a link referring to Wireshark documentation", as you suggested)?

As a first blog post on the use of Wireshark to analyze gRPC traffic I can think of two main alternative scopes:

(1) Only introduce Wireshark and its capabilities, linking to articles elsewhere for practical "How to use Wireshark with gRPC" instructions.

(2) If no such "how to" articles exist elsewhere, then this blog post could be a first one. What I find missing from your current writeup is details for setting up and capturing gRPC network traffic, but I have no idea how easy it would be to cover that concisely -- and maybe that could be the topic of another post.

I'm not saying that you need to change the scope of your blog post (e.g., by adding details about how to capture traffic). But I would like the scope to be clear so that readers know what they're going to learn, and where to go to to learn more.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  name: Huang Qiangxiong
+---
+
+[Wireshark](https://www.wireshark.org/) is the most famous open source network protocol analyzer, which is widely used in network troubleshooting, analysis, software and communications protocol development, and education. Here is the history of Wireshark Protobuf and gRPC dissectors:
+
+- Wireshark 2.6.0 - adds the initial version of Protobuf and gRPC dissectors.
+
+- Wireshark 3.2.0 - supports dissecting serialized Protobuf data (such as gRPC) more precisely according to *.proto files, and supports dissecting gRPC messages of streaming calls.
+
+- Wireshark 3.3.0 - fixes some bugs about parsing *.proto files and supports some new features about Protobuf dissector.
```
Comment (chalin) [inline content/blog/wireshark-protobuf-and-grpc-dissectors.md:15]: An excerpt of a changelog isn't really appropriate. What's your _intent_ behind the inclusion of these log entries? What is the essence of the information that you are trying to conveny?

Comment (ejona86) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md:15]: I think the content here is actually quite helpful, as Arch has Wireshark 3.3, Ubuntu 20.04LTS has Wireshark 3.2, Ubuntu 18.04LTS has 2.4, Debian stable is on 2.6. Users following this will get quite a different experience depending on their version.

Now there is probably a better way to preface this, so it flows and is more clear what the purpose is and the key take aways.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+- You can register your own subdissectors in the 'protobuf_field' dissector table, which is keyed with the full names of the fields, for further parsing Protobuf fields of BYTES or STRING type.
+
+- You can add the mapping record between UDP port and Protobuf message type in the Protobuf protocol preferences dialog, so that the Wireshark can decode the traffic on that UDP port as this Protobuf message.
+
+- You can create your own dissectors to parsing the protocols based on Protobuf over UDP or TCP.
+
+## Examples about using Protobuf and gRPC dissectors
+
+There are some examples to show how to use the Wireshark Protobuf and gRPC dissectors. You can get more details about these examples from [Protobuf wiki](https://gitlab.com/wireshark/wireshark/-/wikis/Protobuf) and [gRPC wiki](https://gitlab.com/wireshark/wireshark/-/wikis/gRPC) pages from Wireshark official website.
```
Comment (ejona86) [inline content/blog/wireshark-protobuf-and-grpc-dissectors.md:37]: Probably want to include "page" within the link itself; as-is it is very easy to think those are protobuf and grpc wikis as opposed to pages on the wireshark wiki.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+![wireshark_protobuf_search_paths](/img/wireshark_protobuf_search_paths.png)
+This preferences dialog can be found by menu 'Edit->Preferences->Protocols->ProtoBuf->Protobuf search paths'.
+
+### gRPC Sample Capture files
+
+There are two gRPC sample capture files [grpc_person_search_protobuf_with_image.pcapng](https://gitlab.com/wireshark/wireshark/-/wikis/uploads/f6fcdceb0248669c0b057bd15d45ab6f/grpc_person_search_protobuf_with_image.pcapng) and [grpc_person_search_json_with_image.pcapng](https://gitlab.com/wireshark/wireshark/-/wikis/uploads/88c03db83efb2e3253c88f853d40477b/grpc_person_search_json_with_image.pcapng) on the [SampleCaptures page](https://gitlab.com/wireshark/wireshark/-/wikis/SampleCaptures) of Wireshark. The difference between the two captures is that the message of former is serialized in Protobuf, and the latter is serialized in JSON.
```
Comment (ejona86) [inline content/blog/wireshark-protobuf-and-grpc-dissectors.md:107]: Nice. Sample captures make it a lot easier for users to follow along.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+#### gRPC Capture serialized in JSON
+
+Following is a screenshot of [grpc_person_search_json_with_image.pcapng](https://gitlab.com/wireshark/wireshark/-/wikis/uploads/88c03db83efb2e3253c88f853d40477b/grpc_person_search_json_with_image.pcapng), which serializes gRPC message with JSON:
+![wireshark_grpc_json_search_response](/img/wireshark_grpc_json_search_response.png)
+
+>Note, if the gRPC message is serialized in JSON, the *'Protobuf Search Paths'* does not need to be configured.
+
+### Protobuf Sample Capture files
+
+#### Built-in Protobuf UDP dissector
```
Comment (ejona86) [inline content/blog/wireshark-protobuf-and-grpc-dissectors.md:140]: I don't think this UDP and custom dissectors fits all that well in this post. It is substantially more complicated and I'm worried that it will scare off users thinking the normal flow is complex. The large lua snippet also doesn't seem all that well suited for a blog post. (I can tell it isn't actually _that_ complicated, but it seems complex and is more than users need.)

That said, it looks like it is good content. So if this content isn't elsewhere, it seems it should find a home. There's a small chance it could make sense to have a "Advanced Protobuf" themed post talking about this, since Protobuf is important to the gRPC community. But raw UDP and TCP aren't all that important to the gRPC community, so it seems a weak fit.

**Review by ejona86 — COMMENTED** [review]
"Protobuf" is fine to use, as we do use it constantly in speak and writing. It's actually pretty easy to forget that they have a fuller name. But we should probably ease into it and refer to "Protocol Buffers" initially in the title and first sentence.

The content itself looks pretty good, but it seems there will need to be a reasonable amount of editorial changes to improve flow.

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
+---
+title: Wireshark Protobuf and gRPC dissectors
+date: 2020-10-20
+spelling: cSpell:ignore Huang Qiangxiong
+author:
+  name: Huang Qiangxiong
```
[the remark this answers is not in the document]
Comment (huangqiangxiong) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md]: done

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md (hunk 2 above)
Comment (huangqiangxiong) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md]: I have moved the history to a section near the end. How about that?

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md (hunk 3 above)
Comment (huangqiangxiong) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md]: done

File: content/blog/wireshark-protobuf-and-grpc-dissectors.md (hunk 5 above)
Comment (huangqiangxiong) [reply content/blog/wireshark-protobuf-and-grpc-dissectors.md]: I have removed these sections, and instead leaving a link in "Read more about Protobuf Dissector" section.

**Comment by huangqiangxiong** [comment]
@chalin @ejona86, I have updated and committed a new version according to your comments. Please help to review, thanks.

**Comment by chalin** [comment]
Thanks for the updates! I'll get back to you as soon as I can.

**Comment by chalin** [comment]
@huangqiangxiong - could you rebase and (force) push? I'd be ready soon to make a pass over this PR.

File: content/blog/how-to-analyze-grpc-with-wireshark.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
+---
+title: How to analyze gRPC with Wireshark
+date: 2020-10-20
```
Comment (chalin) [inline content/blog/how-to-analyze-grpc-with-wireshark.md:3]: While you're making updates/rebasing, could you bring this date closer to today; say, 2020-11-27?

File: content/blog/how-to-analyze-grpc-with-wireshark.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+author:
+  name: Huang Qiangxiong
+  link: https://github.com/huangqiangxiong
+---
+
+[Wireshark](https://www.wireshark.org/) is the most famous open source network protocol analyzer, which is widely used in network troubleshooting, analysis, software and communications protocol development, and education. With Wireshark, people can analyze the messages of gRPC protocol that transferred over the wire (network), or learning the binary wire format of these protocols. This tutorial will show you how to configure and use Wireshark gRPC dissector to analyze gRPC with *.proto files. Note that the gRPC dissector actually relies on the Protocol Buffers (Protobuf) dissector, and some functions (including setting *.proto files paths) are accomplished through the Protobuf dissector. Wireshark started to support parsing gRPC (and Protobuf) in version 2.6.0, and some new features (including supporting *.proto* files) have been added in later versions. You can refer to section [History of Wireshark gRPC and Protobuf Dissectors](#History of Wireshark gRPC and Protobuf Dissectors) for details. Wireshark 3.4.0 is the latest stable version, is also the version we used to introduce examples later.
```
Comment (chalin) [inline content/blog/how-to-analyze-grpc-with-wireshark.md:10]: If you look at the preview, you'll see that this paragraph doesn't render correctly, in particular, because of this link:

    [History of Wireshark gRPC and Protobuf Dissectors](#History of Wireshark gRPC and Protobuf Dissectors)

This is the proper anchor name: `#history-of-wireshark-grpc-and-protobuf-dissectors`

File: content/blog/how-to-analyze-grpc-with-wireshark.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+## Capture gRPC traffic
+
+There are many ways to capture the network traffic of a gRPC conversation:
+
+- Windows: Use Wireshark to capture traffic from many different network media types, including Ethernet, Wireless LAN, and more.
+
+- Linux/UNIX: Use Tcpdump.
+
+Please refer to [Wireshark User's Guide]https://www.wireshark.org/docs/wsug_html_chunked/ for how to capture network packets files that can be recognized by Wireshark.
```
Comment (chalin) [inline content/blog/how-to-analyze-grpc-with-wireshark.md:38]: There's an improperly formatted link here.

File: content/blog/how-to-analyze-grpc-with-wireshark.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+Note that now only the capture files that sending gRPC messages in plaintext mode can be parsed by Wireshark. For example, you have to setup a gRPC client channel using plaintext mode in java like:
+```java
+NettyChannelBuilder.forAddress(host, port).negotiationType(NegotiationType.PLAINTEXT).build();
+```
+
+In theory, as long as the session secret keys of TLS can be exported in format of Wireshark [key log file](https://gitlab.com/wireshark/wireshark/-/wikis/tls) from client or server side, the gRPC traffic encrypted in TLS can also be parsed by Wireshark. But as I know, gRPC stub libraries do not support exporting session secret keys of TLS now or not support directly. Hope this feature might be supported in the future.
```
Comment (chalin) [inline content/blog/how-to-analyze-grpc-with-wireshark.md:45]: @ejona86 @mjpitz @jtattermusch - given the limitations mentioned in lines 40-45 (copied below for your convenience), how useful is this really? (Emphasis mine)

> Note that now **only** the capture files that sending gRPC messages in **plaintext mode** can be parsed by Wireshark. ...
>
> **In theory**, as long as the session secret keys of TLS can be exported in format of Wireshark [key log file](https://gitlab.com/wireshark/wireshark/-/wikis/tls) from client or server side, the gRPC traffic encrypted in TLS can also be parsed by Wireshark. But as I know, **gRPC stub libraries do not support exporting session secret keys** of TLS now or not support directly. Hope this feature might be supported in the future.

So, in theory, this is usable, but in practice it isn't 😄 ?

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 7 above)
Comment (huangqiangxiong) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: done

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 8 above)
Comment (huangqiangxiong) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: done

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 10 above)
Comment (huangqiangxiong) [reply content/blog/how-to-analyze-grpc-with-wireshark.md:45]: Netty supports to export session secret Keys of tls https://github.com/netty/netty/blob/master/handler/src/main/java/io/netty/handler/ssl/SslMasterKeyHandler.java as wireshark key log file format.
But it seems grpc-java does not merge it yet.

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 9 above)
Comment (huangqiangxiong) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: done

**Comment by chalin** [comment]
Thanks for the updates.

> @huangqiangxiong - could you rebase and (force) push? I'd be ready soon to make a pass over this PR.

Have you been able to rebase and push?

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 10 above)
Comment (chalin) [reply content/blog/how-to-analyze-grpc-with-wireshark.md:45]: Thanks for the clarifications. I'll wait for @ejona86 @mjpitz @jtattermusch to comment, which most likely won't be until next week.

**Comment by chalin** [comment]
Oups, I committed some copyedits directly to this PR via https://github.com/grpc/grpc.io/pull/467/commits/9a456a86c76b1a140da39011ac51d707c0f94df9 (see above). That wasn't my intention; what I _wanted_ to do was submit a PR _over_ your PR so that you could then review, approve and then merge my changes back into this PR. This was meant to be a practice run. I'll try to do differently next time.

I'm mentioning all of this just because **you'll need to pull my changes in before you rebase**.

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 10 above)
Comment (huangqiangxiong) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: I find there might have been some progress on it (https://github.com/grpc/grpc-java/issues/7199).

**Comment by huangqiangxiong** [comment]
> Oups, I committed some copyedits directly to this PR via [9a456a8](https://github.com/grpc/grpc.io/commit/9a456a86c76b1a140da39011ac51d707c0f94df9) (see above). That wasn't my intention; what I _wanted_ to do was submit a PR _over_ your PR so that you could then review, approve and then merge my changes back into this PR. This was meant to be a practice run. I'll try to do differently next time.
> 
> I'm mentioning all of this just because **you'll need to pull my changes in before you rebase**.

@chalin I tried to find the 'rebase' button on the web UI, but I couldn't find it. Instead, I have run the following commands (as I usually do on gitlab):
- git clone https://github.com/huangqiangxiong/grpc.io.git
- git remote add upstream https://github.com/grpc/grpc.io.git
- git checkout wireshark-protobuf-and-grpc-dissectors
- git pull --rebase upstream main
- git push origin +HEAD

Is this the correct way to rebase and force push? But it shows some checks were not successful now.

**Comment by chalin** [comment]
@huangqiangxiong - your force push seems to have worked, thanks! I cleared the preview-server caches, and now the previews are being successfully built. I'll continue my review and submit some edits as soon as I can.

File: content/blog/how-to-analyze-grpc-with-wireshark.md (hunk 10 above)
Comment (ejona86) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: Plaintext is (unfortunately) pretty common and many times when using Wireshark you can make it plaintext when you need to. I do think plain-text is enough for the blog post. This raises awareness that things like proto-inspection are even available, and common paths can become smoother.

SSLKEYLOGFILE may be functional today in C; it doesn't appear to be in Java, although I didn't investigate why. Both use boringssl which seems to have support; maybe we need a special flag. It looks there's a way to get the keys today [with one line of Go TLS configuration](https://golang.org/pkg/crypto/tls/#example_Config_keyLogWriter).

Many people aren't using Wireshark, so nobody's worked to make keys available. If we say "no point in using Wireshark because keys are available" then it becomes a chicken and egg problem.

Comment (chalin) [reply content/blog/how-to-analyze-grpc-with-wireshark.md]: That makes sense. Thanks for the thoughtful feedback @ejona86.

**Comment by chalin** [comment]
>  I wanted to do was submit a PR over your PR 

@huangqiangxiong - I haven't been able to find a way to do this for your PR (although I've seen it work before), so this is what I've done/decided to do:

- I've marked the post as a draft, that way it won't show up on the production (or even the preview) server yet.
- I'll accept and merge this PR.
- Finally, I'll submit changes to your blog post as separate PR(s) for your approval.

Once we're both satisfied with the post, we can drop the draft status and it will appear on the production server.

I'm hoping that everyone is ok with all this!

**Comment by chalin** [comment]
Thanks for you patience @huangqiangxiong. In the end we'll have a nicely polished blog post!

**Comment by huangqiangxiong** [comment]
> > I wanted to do was submit a PR over your PR
> 
> @huangqiangxiong - I haven't been able to find a way to do this for your PR (although I've seen it work before), so this is what I've done/decided to do:
> 
> * I've marked the post as a draft, that way it won't show up on the production (or even the preview) server yet.
> * I'll accept and merge this PR.
> * Finally, I'll submit changes to your blog post as separate PR(s) for your approval.
> 
> Once we're both satisfied with the post, we can drop the draft status and it will appear on the production server.
> 
> I'm hoping that everyone is ok with all this!

OK. Thanks!
